test/data/s3/s3_data.py

```diff
@@ -0,0 +1,451 @@
+
+def ensure_test_data():
+    # update S3ROOT in __init__.py to get a fresh set of data
+    print('Ensuring that test data exists at %s' % S3ROOT)
+    mark = urlparse(os.path.join(S3ROOT, 'ALL_OK'))
+    try:
+        # Check if parquet data exists and has been modified in the last
+        # 29 days (default TTL - 1).
+        import datetime
+        today = datetime.date.today()
+        delta = datetime.timedelta(days=29)
+        s3client.head_object(Bucket=mark.netloc, Key=mark.path.lstrip('/'),
+                             IfModifiedSince=str(today - delta))
+        print('All data ok.')
+    except:
+        print('Uploading test data')
+        def _do_upload(prefix, filespecs, meta=None):
+            for fname, size in filespecs.items():
+                if size is not None:
+                    f = RandomFile(prefix, fname, size)
+                    url = urlparse(f.url)
+                    # For metadata, we don't actually touch RandomFile
+                    # (since it is the same) but we modify the path to post-pend
+                    # the name
+                    print('Case %s: %s started' % (prefix, f.url))
+                    s3client.upload_fileobj(f.fileobj(),
+                                            url.netloc,
+                                            url.path.lstrip('/'))
+                    print('Case %s: %s added' % (prefix, f.url))
+                    if meta is not None:
+                        for metaname, metainfo in meta.items():
+                            new_url = "%s_%s" % (f.url, metaname)
+                            url = urlparse(new_url)
+                            print('Case %s: %s started' % (prefix, new_url))
+                            extra = {}
+                            content_type, user_meta = metainfo
+                            if content_type:
+                                extra['ContentType'] = content_type
+                            if user_meta:
+                                new_meta = {
+                                    'metaflow-user-attributes': json.dumps(user_meta)}
+                                extra['Metadata'] = new_meta
+                            s3client.upload_fileobj(f.fileobj(),
+                                                    url.netloc,
+                                                    url.path.lstrip('/'),
+                                                    ExtraArgs=extra)
+                            print('Case %s: %s added' % (prefix, new_url))
+
+        for prefix, filespecs in BIG_DATA + FAKE_RUN_DATA:
+            _do_upload(prefix, filespecs)
+        for prefix, filespecs in BASIC_DATA:
+            _do_upload(prefix, filespecs, meta=BASIC_METADATA)
+
+        s3client.upload_fileobj(to_fileobj('ok'),
+                                Bucket=mark.netloc,
+                                Key=mark.path.lstrip('/'))
+        print('Test data uploaded ok')
+
+ensure_test_data()
```

Comment:
Is there an easy way to call this inside a fixture or something like that? I'd generally avoid doing any non trivial work on module import. However, as this is just test code, I don't feel very strongly

Reply:
I'm not sure. As you say, it's a test so I didn't look to change it much (I just copied what we have). If it bugs you, definitely propose something else :).